### PR TITLE
feat: cylindrical detector from DD4hep

### PR DESCRIPTION
**Pull request by asalzburger**
This PR introduces the possibility to build cylindrical detectors via a blue print definition from `DD4hep`.

It mainly adds two classes:
- `DD4hepBlueprint` which parses the DD4hep Detector Element tree and creates a blueprint
- `DD4hepDetectorStructure` which can turn this into a detector

Working with the Blueprint allows to get rid of the `DD4hepVolumeStructure` as this is part of the blueprint definition.

For debugging, is also refines the drawing of the `Blueprint` to `.dot` which is now decoupled from the class into a helper function.

It adds an extensive test on a simple detector,  and to make sure it does what it should do, I introduced a new 
- `DD4hepDiscLayerStructureTests` unit test.

Screenshot 2023-10-25 at 17 00 06

The building seems to work, navigation is not tested yet on this one.

Screenshot 2023-10-25 at 17 01 31

### CODE REVIEW

File: Core/include/Acts/Detector/detail/BlueprintDrawer.hpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  Node internals = Node{"doubleoctagon", "cadetblue1"};
+  Node geoID = Node{"box", "azure"};
+  Node roots = Node{"box", "darkkhaki"};
+};
+
+/// @brief Geneate the shape string
+/// @param s the shape of the object
+/// @param c the color of the object
+/// @return a string with the shape and color
+std::string shapeStr(const Options::Node& node) {
```
Comment (paulgessinger) [inline Core/include/Acts/Detector/detail/BlueprintDrawer.hpp:58]: Not super critical, but these could go into a `.cpp` file.

File: Core/include/Acts/Detector/detail/BlueprintDrawer.hpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+/// @brief  Turn into a dot output
+///
+/// @tparam the stream type to be used
+///
+/// @param ss the stream into which the dot output should be written
+/// @param node the node to be drawn
+/// @param options the options for the drawer
+
+template <typename stream_type>
```
Comment (paulgessinger) [inline Core/include/Acts/Detector/detail/BlueprintDrawer.hpp:100]: Why not concrete `std::ostream`?

File: Core/include/Acts/Utilities/StringHelpers.hpp [hunk 3]
```diff
@@ -103,4 +104,22 @@ inline std::string toString(const Acts::Transform3& transform,
   return sout.str();
 }
 
+/// Print out a vector of ActsScalar
+/// @param pVector The vector to print
+/// @param precision Numeric output precision
+inline std::string toString(const std::vector<ActsScalar>& pVector,
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/StringHelpers.hpp:110]: `.cpp`?

File: Core/include/Acts/Geometry/VolumeBounds.hpp [hunk 4]
```diff
@@ -66,6 +66,9 @@ class VolumeBounds {
     eOther = 6
   };
 
+  /// Static member to get the name of the BoundsType
+  static const std::vector<std::string> s_boundsTypeNames;
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/VolumeBounds.hpp:70]: `VolumeBounds` is already a base class with virtual methods. How about we add this ass a method derived classes just implement?

File: Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepBinningHelpers.hpp [hunk 5]
```diff
@@ -30,6 +30,28 @@ namespace Acts {
 static std::vector<std::tuple<std::string, BinningValue>> allowedBinnings = {
     {"x", binX}, {"y", binY}, {"z", binZ}, {"phi", binPhi}, {"r", binR}};
 
+/// Helper method to convert the string to binning value
+///
+/// @param binningString
+///
+/// @return a binningValue
+inline BinningValue stringToBinningValue(const std::string &binningString) {
```
Comment (paulgessinger) [inline Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepBinningHelpers.hpp:38]: I feel like we have a helper like this somewhere already, but I don't remember where.

File: Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+      <argument value="acts_container_binning: str = z"/>
+    </plugin>	
+    <plugin name="DD4hep_ParametersPlugin">
+      <argument value="PixelNegativeEndcap"/>
+      <argument value="acts_container: bool = true"/>
+      <argument value="acts_container_type: int = 3"/>
+      <argument value="acts_container_bvalues_n: int = 3"/>
+      <argument value="acts_container_bvalues_0: double = 20*mm"/>
+      <argument value="acts_container_bvalues_1: double = 150*mm"/>
+      <argument value="acts_container_bvalues_2: double = 200*mm"/>
```
Comment (paulgessinger) [inline Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp:111]: Do you think it's be worth to add vector support to the parameter plugin on the DD4hep side to try to make this simpler?

**Review by paulgessinger — COMMENTED** [review]
Pretty good. A number of functions I think could move to corresponding `.cpp` files, otherwise this looks good to me.

File: Core/include/Acts/Detector/detail/BlueprintDrawer.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/BlueprintDrawer.hpp:100]: Changed.

File: Core/include/Acts/Detector/detail/BlueprintDrawer.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/BlueprintDrawer.hpp:58]: done.

File: Core/include/Acts/Geometry/VolumeBounds.hpp (hunk 4 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/VolumeBounds.hpp:70]: I need access to the collection of names here, so I bundled them in the base class - this is for readability of some output file only.

File: Core/include/Acts/Utilities/StringHelpers.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/StringHelpers.hpp:110]: Done.

File: Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepBinningHelpers.hpp (hunk 5 above)
Comment (asalzburger) [reply Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepBinningHelpers.hpp:38]: We had (in the JSON writing), but I removed this in favour in nlohmann's auto-generation of this for enums.

File: Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp (hunk 6 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp:111]: I have worked around it now, but yes, `vector<double>` and `vector<int>` is something I could use there.

File: Core/src/Geometry/VolumeBounds.cpp [hunk 7]
```diff
@@ -8,6 +8,10 @@
 
 #include "Acts/Geometry/VolumeBounds.hpp"
 
+const std::vector<std::string> Acts::VolumeBounds::s_boundsTypeNames = {
+    "Cone",          "Cuboid",    "CutoutCylinder", "Cylinder",
+    "GenericCuboid", "Trapezoid", "Other"};
+
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/src/Geometry/VolumeBounds.cpp:14]: That's actually what clang-format came up with ...

**Comment by asalzburger** [comment]
Hi @paulgessinger @AJPfleger  - I think I have addressed all your comments.

File: Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp [hunk 8]
```diff
@@ -318,11 +318,12 @@ BOOST_AUTO_TEST_CASE(DD4hepCylidricalDetectorExplicit) {
...
   auto dd4hepBlueprint = bp.create(bpCache, world);
 
-  // We should have 5 store entries now
+  // We should have 6 store entries now
+  // 1 : beam pipe (empty)
   // 1 : endcap
   // 2 : barrel
   // 1 : endcap
-  BOOST_TEST(bpCache.dd4hepStore.size(), 5u);
+  BOOST_CHECK(bpCache.dd4hepStore.size() == 6u);
```
Comment (paulgessinger) [inline Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp:326]: Not critical, but you could do 

```suggestion
  BOOST_CHECK_EQUAL(bpCache.dd4hepStore.size(), 6u);
```

which gives you separate printouts of the arguments in the error case.

**Review by paulgessinger — APPROVED** [review]

File: Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp (hunk 8 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/DD4hep/DD4hepCylindricalDetectorTests.cpp:326]: Ok, I see - I can do a cleanup of those throughout then.

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [d4842c7ac443a63753e8460d2bda4b2c5759ee89]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6409318)!

|status|job|
|------|---|
|:red_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33534369) |
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532512) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532511) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532510) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532509) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532508) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532507) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532506) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532505) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532504) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532503) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532502) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532501) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532500) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532499) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532498) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532496) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33532495) |

**Comment by paulgessinger** [comment]
Failure is due to metadata check.

**Pull request merged**